File: scripts/gateway.test.mjs
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { renderGatewayConf } from "./render-gateway.mjs";
import { parseSanOutput, sanEntry } from "./gen-certs.mjs";

const TEMPLATE = ["server {", "@@HTTP_LISTEN@@", "@@TLS_LISTEN@@", "}"].join("\n");

describe("renderGatewayConf", () => {
  it("renders plain HTTP byte-identically (no gaps)", () => {
    const out = renderGatewayConf(TEMPLATE, {});
    assert.equal(out, ["server {", "    listen 80;", "}"].join("\n"));
  });

  it("adds the 443 listener and drops the 80 listener on redirect", () => {
    const out = renderGatewayConf(TEMPLATE, { TLS_HOSTS: "workbench.lan", TLS_REDIRECT: "1" });
    assert.ok(out.includes("listen 443 ssl;"));
    assert.ok(!out.includes("listen 80;"));
  });
});

describe("sanEntry", () => {
  it("maps IPs to IP entries and names to DNS", () => {
    assert.equal(sanEntry("10.0.0.63"), "IP:10.0.0.63");
    assert.equal(sanEntry("::1"), "IP:::1");
    assert.equal(sanEntry("workbench.lan"), "DNS:workbench.lan");
  });
});

describe("parseSanOutput", () => {
  it("parses openssl subjectAltName output into comparable tokens", () => {
    const tokens = parseSanOutput(
      "X509v3 Subject Alternative Name: \n    DNS:workbench.lan, DNS:localhost, IP Address:127.0.0.1\n",
    );
    assert.deepEqual(tokens, ["DNS:workbench.lan", "DNS:localhost", "IP Address:127.0.0.1"]);
  });
});

File: scripts/gateway_test.py
"""Unit tests for render_gateway.py and gen_certs helpers.

Run: python3 -m unittest discover -s scripts -t . -p "*_test.py".
"""

from __future__ import annotations

import unittest

from scripts.gen_certs import parse_san_output, san_entry
from scripts.render_gateway import render_gateway_conf

TEMPLATE = "server {\n@@HTTP_LISTEN@@\n@@TLS_LISTEN@@\n}"


class RenderGatewayConfTest(unittest.TestCase):
    def test_plain_http_byte_identical_no_gaps(self) -> None:
        out = render_gateway_conf(TEMPLATE, {})
        self.assertEqual(out, "server {\n    listen 80;\n}")

    def test_redirect_adds_443_and_drops_80(self) -> None:
        out = render_gateway_conf(
            TEMPLATE, {"TLS_HOSTS": "workbench.lan", "TLS_REDIRECT": "1"}
        )
        self.assertIn("listen 443 ssl;", out)
        self.assertNotIn("listen 80;", out)

    def test_unknown_tokens_pass_through(self) -> None:
        out = render_gateway_conf("x\n@@NOPE@@\ny", {})
        self.assertEqual(out, "x\n@@NOPE@@\ny")


class SanEntryTest(unittest.TestCase):
    def test_maps_ips_and_names(self) -> None:
        self.assertEqual(san_entry("10.0.0.63"), "IP:10.0.0.63")
        self.assertEqual(san_entry("::1"), "IP:::1")
        self.assertEqual(san_entry("workbench.lan"), "DNS:workbench.lan")


class ParseSanOutputTest(unittest.TestCase):
    def test_parses_openssl_san_output(self) -> None:
        tokens = parse_san_output(
            "X509v3 Subject Alternative Name: \n    DNS:workbench.lan, DNS:localhost, IP Address:127.0.0.1\n"
        )
        self.assertEqual(
            tokens, ["DNS:workbench.lan", "DNS:localhost", "IP Address:127.0.0.1"]
        )


if __name__ == "__main__":
    unittest.main()
